### PR TITLE
[SPARK-18057][SS] Update Kafka client version from 0.10.0.1 to 2.0.0

### PR DESCRIPTION
**Pull request by tedyu**
## What changes were proposed in this pull request?

This PR upgrades to the Kafka 2.0.0 release where KIP-266 is integrated.

## How was this patch tested?

This PR uses existing Kafka related unit tests

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)
(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaTestUtils.scala [hunk 1]
```diff
@@ -96,10 +101,13 @@ class KafkaTestUtils(withBrokerProps: Map[String, Object] = Map.empty) extends L
...
     // Zookeeper server startup
-    zookeeper = new EmbeddedZookeeper(s"$zkHost:$zkPort")
+    val zkSvr = s"$zkHost:$zkPort";
+    zookeeper = new EmbeddedZookeeper(zkSvr)
     // Get the actual zookeeper binding port
     zkPort = zookeeper.actualPort
-    zkUtils = ZkUtils(s"$zkHost:$zkPort", zkSessionTimeout, zkConnectionTimeout, false)
+    zkUtils = ZkUtils(zkSvr, zkSessionTimeout, zkConnectionTimeout, false)
+    zkClient = KafkaZkClient(zkSvr, false, 6000, 10000, Int.MaxValue, Time.SYSTEM)
+    adminZkClient = new AdminZkClient(zkClient)
```
Comment (ijuma) [inline external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaTestUtils.scala]: Can we use the Java AdminClient instead of these internal classes?

Comment (tedyu) [reply external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaTestUtils.scala]: AdminClient is abstract.
KafkaAdminClient doesn't provide addPartitions.

Mind giving some pointer ?

Comment (ijuma) [reply external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaTestUtils.scala]: AdminClient.create gives you a concrete instance. createPartitions is the method you're looking for.

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaTestUtils.scala [hunk 2]
```diff
@@ -203,7 +215,13 @@ class KafkaTestUtils(withBrokerProps: Map[String, Object] = Map.empty) extends L
 
   /** Add new partitions to a Kafka topic */
   def addPartitions(topic: String, partitions: Int): Unit = {
-    AdminUtils.addPartitions(zkUtils, topic, partitions)
+    val existingAssignment = zkClient.getReplicaAssignmentForTopics(
+      collection.immutable.Set(topic)).map {
+        case (topicPartition, replicas) => topicPartition.partition -> replicas
+    }
```
Comment (ijuma) [inline external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaTestUtils.scala]: We can get replica assignment information via AdminClient too. I think we should try to avoid the internal `ZkUtils` and `KafkaZkClient` as much as we can.

**Comment by tedyu** [comment]
Currently I am trying to get test suite pass first.

**Comment by tedyu** [comment]
I tried the following change but didn't seem to get more output from Kafka:
```
diff --git a/external/kafka-0-10-sql/src/test/resources/log4j.properties b/external/kafka-0-10-sql/src/test/resources/log4j.properties
index 75e3b53..0d65339 100644
--- a/external/kafka-0-10-sql/src/test/resources/log4j.properties
+++ b/external/kafka-0-10-sql/src/test/resources/log4j.properties
@@ -25,4 +25,4 @@ log4j.appender.file.layout.ConversionPattern=%d{yy/MM/dd HH:mm:ss.SSS} %t %p %c{

 # Ignore messages below warning level from Jetty, because it's a bit verbose
 log4j.logger.org.spark-project.jetty=WARN
-
+log4j.logger.org.apache.kafka=DEBUG
```

File: external/kafka-0-10-sql/pom.xml [hunk 3]
```diff
@@ -29,7 +29,7 @@
   <artifactId>spark-sql-kafka-0-10_2.11</artifactId>
   <properties>
     <sbt.project.name>sql-kafka-0-10</sbt.project.name>
-    <kafka.version>0.10.0.1</kafka.version>
+    <kafka.version>2.0.0-SNAPSHOT</kafka.version>
   </properties>
   <packaging>jar</packaging>
   <name>Kafka 0.10 Source for Structured Streaming</name>
```
Comment (eric-maynard) [inline external/kafka-0-10-sql/pom.xml]: We should change this line to reflect the change too

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaTestUtils.scala (hunk 2 above)
Comment (guozhangwang) [reply external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaTestUtils.scala]: +1

**Comment by tedyu** [comment]
There is only target/surefire-reports/TEST-org.apache.spark.sql.kafka010.KafkaMicroBatchV2SourceSuite.xml under target/surefire-reports

That file doesn't contain test output.

**Comment by tedyu** [comment]
Made some progress in testing.
Now facing:
```
- assign from latest offsets (failOnDataLoss: true) *** FAILED ***
  java.lang.IllegalArgumentException: requirement failed
  at scala.Predef$.require(Predef.scala:212)
  at org.apache.spark.sql.kafka010.KafkaSourceSuiteBase.org$apache$spark$sql$kafka010$KafkaSourceSuiteBase$$testFromLatestOffsets(KafkaMicroBatchSourceSuite.scala:993)
  at org.apache.spark.sql.kafka010.KafkaSourceSuiteBase$$anonfun$37$$anonfun$apply$2.apply$mcV$sp(KafkaMicroBatchSourceSuite.scala:734)
  at org.apache.spark.sql.kafka010.KafkaSourceSuiteBase$$anonfun$37$$anonfun$apply$2.apply(KafkaMicroBatchSourceSuite.scala:732)
  at org.apache.spark.sql.kafka010.KafkaSourceSuiteBase$$anonfun$37$$anonfun$apply$2.apply(KafkaMicroBatchSourceSuite.scala:732)
  at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
```

**Comment by tedyu** [comment]
Located the test output:
```
-rw-r--r-- 1 hbase hadoop 35335485506 Jun 13 20:36 target/unit-tests.log
```
Still need to find out cause for assertion failure.

**Comment by ijuma** [comment]
Any luck getting to the bottom of the issue? It would be great to include this in the next version of Spark?

**Comment by zsxwing** [comment]
@tedyu could you please just bump to 1.1.0, the current official latest release from Apache Kafka?

File: external/kafka-0-10-sql/pom.xml [hunk 4]
```diff
@@ -74,6 +74,11 @@
       <version>${kafka.version}</version>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>org.eclipse.jetty</groupId>
```
Comment (zsxwing) [inline external/kafka-0-10-sql/pom.xml]: Where does this come from? Or it can be just a test dependency?

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaOffsetReader.scala [hunk 5]
```diff
@@ -115,7 +116,7 @@ private[kafka010] class KafkaOffsetReader(
   def fetchTopicPartitions(): Set[TopicPartition] = runUninterruptibly {
     assert(Thread.currentThread().isInstanceOf[UninterruptibleThread])
     // Poll to get the latest assigned partitions
-    consumer.poll(0)
+    consumer.poll(JDuration.ofMillis(0))
```
Comment (zsxwing) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaOffsetReader.scala]: Could you revert these changes? We don't use java.time.Duration in Spark.

**Comment by tedyu** [comment]
w.r.t. stable Kafka release, it seems 2.0.0 RC2 would pass:

http://search-hadoop.com/m/Kafka/uyzND1ClBEezundG1?subj=Re+VOTE+2+0+0+RC2

**Comment by tedyu** [comment]
@ijuma 
Sorry for the late response. 9 days ago I was in China where access to gmail is intermittent.

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaOffsetReader.scala (hunk 5 above)
Comment (tedyu) [reply external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaOffsetReader.scala]: Depending on the Kafka release we agree upon, I can revert.
Duration is recommended API for 2.0.0 release

**Comment by zsxwing** [comment]
@tedyu We don't need to block on the release. I would expect the Kafka client is backward complicity.  In addition, most of Spark users will not upgrade their Kafka at once after the new version comes out. I'm inclined to just use 1.1.0 in this PR.

**Comment by tedyu** [comment]
Not sure what to do with the following build error which is not caused by the PR:
```
[ERROR] /spark/external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaDataConsumerSuite.scala:71: Symbol 'term org.eclipse' is missing from the classpath.
This symbol is required by 'method org.apache.spark.metrics.MetricsSystem.getServletHandlers'.
Make sure that term eclipse is in your classpath and check for conflicting dependencies with `-Ylog-classpath`.
A full rebuild may help if 'MetricsSystem.class' was compiled against an incompatible version of org.
[ERROR]     val kafkaParams = Map[String, Object](
[ERROR]                                          ^
```

**Comment by zsxwing** [comment]
I see. It should be a test dependency. You don't need to add it into the compile scope.

**Comment by zsxwing** [comment]
@tedyu this patch should fix most of unit tests: https://github.com/zsxwing/spark/commit/c2834e2aeaa979f3c4a3f550404856c2de573463

I'm still thinking about how to fix the stress test. By the way, I'm using SBT right now. If you are using maven, it probably misses some test dependencies and maven doesn't add transitive test dependencies automatically. You can manually add them if you find anything is missing.

**Comment by zsxwing** [comment]
@tedyu my branch https://github.com/zsxwing/spark/commits/review-pr21488 should fix most of unit tests. Could you check and add to your PRs? Several tests are still flaky and ignored. If you have time, could you try to fix them? I will take a look tomorrow as well.

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaOffsetReader.scala (hunk 5 above)
Comment (ijuma) [reply external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaOffsetReader.scala]: @zsxwing Why do you want to support Kafka clients jars from 0.10 to 2.0? Since newer clients jars support older brokers, we recommend people use the latest Kafka clients jar whenever possible.

**Comment by zsxwing** [comment]
Okey. In Kafka 1.1.0, deleting a topic when a Kafka client is running may make the client hang at this line forever: https://github.com/apache/kafka/blob/1.1.0/clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java#L1428

The fix is https://issues.apache.org/jira/browse/KAFKA-6979. Before we upgrade to Kafka 2.0.0, we have to ignore these tests.

**Comment by zsxwing** [comment]
@tedyu could you update the PR title and description to reflect the latest changes, such as `[SPARK-18057][SS] Update Kafka client version from 0.10.0.1 to 1.1.0`? Otherwise, LGTM.

**Comment by zsxwing** [comment]
@tedyu I forgot one place: https://github.com/apache/spark/blob/master/external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaContinuousSourceSuite.scala#L32 Could you also disable it? Thanks!

**Comment by tedyu** [comment]
Ryan:
Thanks for the reminder.

I have disabled that test.

**Comment by ijuma** [comment]
1.1.1 has been released, maybe we can upgrade to that.

**Comment by tedyu** [comment]
Thanks for the reminder, @ijuma 

Updated pom.xml and title accordingly.

**Comment by guozhangwang** [comment]
> 1.1.1 has been released, maybe we can upgrade to that.

+1

**Comment by tedyu** [comment]
Test failure was in Hive test, not related to this PR.

**Comment by tedyu** [comment]
@zsxwing 
Is there anything that needs to be done from my side ?

**Comment by zsxwing** [comment]
@tedyu I just realized topic deletion might be used by some users. This Kafka version will introduce a regression for them. Since 2.0 fixes this issue and is going out soon, how about just wait for 2.0?

**Comment by zsxwing** [comment]
@tedyu fyi, here is the last fix for 2.0.0: https://github.com/zsxwing/spark/commit/7ccc520b540106ca4c2dac2e42834984332d6804

**Comment by tedyu** [comment]
```
22:36:05.028 WARN org.apache.spark.scheduler.TaskSetManager: Lost task 0.0 in stage 16314.0 (TID 39181, localhost, executor driver): java.io.FileNotFoundException: File file:/home/jenkins/workspace/SparkPullRequestBuilder/target/tmp/spark-0bbc239c-37c5-4df2-b86d-e9c7628ceb28/f1=1/f2=1/part-00000-390ac6da-50dc-4d32-ba08-462da1e8a0c2.c000.snappy.parquet does not exist
It is possible the underlying files have been updated. You can explicitly invalidate the cache in Spark by running 'REFRESH TABLE tableName' command in SQL or by recreating the Dataset/DataFrame involved.
	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.org$apache$spark$sql$execution$datasources$FileScanRDD$$anon$$readCurrentFile(FileScanRDD.scala:131)
	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.nextIterator(FileScanRDD.scala:182)
	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.hasNext(FileScanRDD.scala:109)
```
Doesn't seem to be related to PR.

**Comment by tedyu** [comment]
retest this please

**Comment by gatorsmile** [comment]
https://github.com/apache/kafka/releases KAFKA 2.0.0 release is publicly available. We can finish it before the code freeze?

**Comment by srowen** [comment]
I might have missed this in the shuffle here -- is this fully compatible with 0.10.x brokers too?

**Comment by zsxwing** [comment]
> I might have missed this in the shuffle here -- is this fully compatible with 0.10.x brokers too?

Yep. @ijuma could you confirm it?

**Comment by ijuma** [comment]
Yes, Kafka clients 0.10.2 and higher support brokers from 0.10.0 and higher if the protocols being used are available in that version. Since we only changed test code in this PR, upgrading to 2.0.0 is safe.

If Spark were to use the AdminClient create topics, then the brokers would have to be 0.10.1 or higher. Other admin protocols like config management were added in more recent versions and so on.

**Comment by srowen** [comment]
@ijuma this changes the version of `kakfa-clients` in compile scope, too -- right? looks like it in the POM. That's not just a test change. It's probably OK if the client is backwards-compatible with 0.10.x brokers, which is what this integration currently supports. I guess I'm always nervous about making a major-version change to a user-visible dependency; this is a big one too. But you all may know better whether there is any potential user impact. Would this change break any user code that was previously compiling against the 0.10.x client?

**Comment by ijuma** [comment]
@srowen I meant that the library is compatible, so if you just change the version in the pom, it's fine. If you changed the code to use some methods in AdminClient, then you'd have to be more careful. Does that make sense?

The major version bump is because we now require Java 8 (which I believe you require too) and the Scala clients were removed.

**Comment by zsxwing** [comment]
@srowen just to be clear, `AdminClient` is not in `kafka-clients` jar. The user has to add `kafka` jar as a dependency to use `AdminClient`.

In addition, even if this upgrade has some unexpected issues, the user can still switch back to an old `kafka-clients` version by pining the version in their pom.xml. This Kafka connector in Spark uses only 0.10 APIs.

**Comment by ijuma** [comment]
Yeah, the Java client libraries have been evolved in a compatible manner for the most part since 0.10.0. The set of broker versions supported by 0.10.0 and 2.0.0 is exactly the same.

The consumer/producer API has been enriched (transactions, idempotent producer, offsetsForTimes), but the existing methods have been kept. A small number of deprecated, but rarely used methods have been removed (not in KafkaProducer or KafkaConsumer though):

https://github.com/apache/kafka/commit/a4c292173618e22b3f99ab8615c6b585c7688a40

In 0.10.1, a heartbeat thread was added to the Java consumer:

https://cwiki.apache.org/confluence/display/KAFKA/KIP-62%3A+Allow+consumer+to+send+heartbeats+from+a+background+thread

This is helpful to users who could not call `poll` often enough, but the default configs should be fine for most cases.

**Comment by ijuma** [comment]
@zsxwing `org.apache.kafka.clients.admin.AdminClient` is in the clients jar. It's not relevant for this PR, I am just mentioning it so if Spark decides to migrate to that eventually. I see code using `ZkUtils`, which is an internal class that will be removed eventually.

**Comment by zsxwing** [comment]
@ijuma I see. I was looking at 0.10 jar. Thanks for correcting me.

**Comment by ijuma** [comment]
Anyway, overall I think you should definitely make this change. Spark users are currently penalised heavily when running on clusters with the message format introduced in 0.11.0, which has important resiliency improvements (KIP-101). And, as @zsxwing said, people can choose an older version if necessary (I can't think of a reason why, the reason why have focused on making client jars compatible is so that people can just use the latest independently of broker versions).

**Comment by srowen** [comment]
Yes, good argument. Having been burned in the past (actually, by Kafka and ZK changes, though that's in the past), I'm aware that even compiling against version B instead of A, with no library code changes, may still mean the library no longer runs against version A. Changing exception signatures is also sometimes a problem.

It may not be the case here, unlikely, but do want to think this through fully. Yes differences like https://github.com/apache/kafka/commit/a4c292173618e22b3f99ab8615c6b585c7688a40 are what I am worried about.

Still I have no concrete example of this type of problem in the wild.

**Comment by zsxwing** [comment]
@srowen May I read your comment as "no objections"? The current PR looks good to me. If you don't have objections, I will go ahead and merge it.

**Comment by srowen** [comment]
Yes no objections. There's no concrete problem, and there is upside to making the change. I think you're aware of the potential issues, and have thought through them in the context of deeper knowledge of Kafka's APIs than I have.

**Comment by zsxwing** [comment]
Thanks! Merging to master.

**Comment by tedyu** [comment]
I used the following command and the test passed:

mvn test -Phadoop-2.6 -Pyarn -Phive -Dtest=KafkaMicroBatchSourceSuite -rf external/kafka-0-10-sql

Please take a look at the 'Disk error' message and see if it was related to test failure.

**Comment by srowen** [comment]
Hm, looking at the dashboard, I don't see this failure consistently in the master tests:

https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Test%20(Dashboard)/

Let's keep a close eye on it to see if it repeats. It might have been something transient.

**Comment by ijuma** [comment]
@wangyum, can you please file a Kafka JIRA with details of what the test is doing (even if the failure is transient)? From the stacktrace, it looks like a potential broker issue (assuming there are no real disk issues where these tests were executed). If there is indeed a new issue (we have to verify since the test seems to be transient), it would likely only affect tests.

**Comment by srowen** [comment]
Ack I missed something here: there's an override of kafka.version for Scala 2.12, from when it had to be bumped _up_ to work with 2.12. That no longer works when compiling with 2.12. I'll submit a follow up.
